### ISSUE TITLE
Support embedding native views in Windows shell

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#31713]
We'll need support for embedding native views in Flutter views (e.g., for embedding a web view, as on mobile).

Obviously this depends on having a view-based Windows shell first (#30726).

**Comment by SahajRana** [comment]
Eagerly waiting for it.

**Comment by wanjm** [comment]
what is the reason that there is no milestone or project to track such issue;

**Comment by stuartmorgan** [comment]
> what is the reason that there is no milestone or project to track such issue;

See https://github.com/flutter/flutter/wiki/Issue-hygiene#when-will-my-bug-be-fixed

**Comment by BrianWatson0921** [comment]
Hello, just checking.. is there any timeline as to when it's expected this may be implemented? Thanks.

**Comment by majdi21** [comment]
Hello,

   Are there any plans to add a windows platform view? I would imagine this would be important for Flutter adoption on that platform. Please share any information you have. Thank you :)

**Comment by vanlooverenkoen** [comment]
With the stable release of windows launched today. When can we expect this to be implemented? Is there a roadmap for platform views on desktop in general?

**Comment by alexmercerind** [comment]
I was able to take out a section of the Flutter window. Allowing click/other interactions through the "cutout" will not be hard (can be done by sending `HTTRANSPARENT` instead of `HTCLIENT` for that particular coordinate zone).

Retrieving global coordinates of this "cutout" is not a difficult task & positioning another window (which we want to "embed" as a "platform view") (with its taskbar entry removed) will result in an interactable natively rendered surface inside Flutter view upon which other Flutter `Widget`s can also be placed.

<img width="623" alt="Screenshot 2022-02-20 161002" src="https://user-images.githubusercontent.com/28951144/154853810-b6f2518e-bb4d-4549-86d9-d13cd026bc21.png">

This will result in:
- A natively rendered webview for Flutter on Windows ([webview_windows](https://pub.dev/packages/webview_windows) is available but uses new `Windows.Graphics.Capture` which is only available on later Windows versions & this setup of Webview Runtime 2 + `Windows.Graphics.Capture` degrades performance a lot).
- [Finally] a hardware accelerated video playback which I couldn't achieve using `dart_vlc`. (#37673)

**Comment by Razzile** [comment]
@alexmercerind Would this allow us to create a native HWND inside a flutter widget which DirectX could then be initialized with? I am interested in combining a Flutter app with a Direct3D renderer

**Comment by alexmercerind** [comment]
> @alexmercerind Would this allow us to create a native HWND inside a flutter widget which DirectX could then be initialized with? I am interested in combining a Flutter app with a Direct3D renderer

Yes, you'll be able to render & interact with a native HWND inside Flutter Widget tree (and other Flutter widgets can be placed on top). Currently I have only tested on Windows 10 & 10+ (maybe `WS_EX_LAYERED` can work on older versions).  This won't require any changes to the engine or framework, so I will likely publish it as a plugin.
I can achieve same results on Linux with some work, however can't confirm that yet.

**Comment by alexmercerind** [comment]
@Razzile

> @alexmercerind Would this allow us to create a native HWND inside a flutter widget which DirectX could then be initialized with? I am interested in combining a Flutter app with a Direct3D renderer

What can be a better app than `notepad.exe` to embed inside Flutter.
I think this will already solve your issue or a video playback for example.
![Code_pDg6rcG21m](https://user-images.githubusercontent.com/28951144/155029550-6d492407-5c39-4148-b7ab-88bf63b8d8b1.gif)
_I randomly placed a FAB on top._

Before anyone says anything, I know there's a delay between positioning & it can be fixed. There are many ways to ensure  that it doesn't happen. It's just hardcoded in an hour & just a proof of concept. I think the primary reason for it to happen is that method channel yields result asynchronously in the lambda (maybe FFI can work better).

**Comment by vanlooverenkoen** [comment]
@alexmercerind is not the hero we deserved, but the hero we needed

**Comment by damywise** [comment]
@eggfly is your image correct? I can't open it. It only shows as a link to this thread.
![image](https://user-images.githubusercontent.com/25608913/159516557-44a324b9-25aa-492e-8efc-0738219e343c.png)

**Comment by stuartmorgan** [comment]
> But I have no idea with letting platform view at the middle of z-order, in the middle of the flutter views

That's what [FlutterCompositor](https://github.com/flutter/engine/blob/6c67716b081a3614305f99b4e6ade2562f974fee/shell/platform/embedder/embedder.h#L1667-L1681) is for.

> 2\. if there are flutter view above the native views, we should make a overlay view and copy these flutter views' pixel data and may also handle these input events manually.

Why?

**Comment by alexmercerind** [comment]
> I made a prototype to embed a webview2 ([Microsoft Edge's webview](https://docs.microsoft.com/en-us/microsoft-edge/webview2/)) along with flutter.
> Demo project: https://github.com/eggfly/flutter_windows_webview2
> 
I'm not sure how this is a step forward, when there is [webview_windows](https://pub.dev/packages/webview_windows) already available (as derived from the conversation at #37597). On the other hand, I don't think `::SetParent`-ing two windows inside same view will result in any good experience (why not launch another window for webview, when you are _fine_ with obstructing the view?). It'll never be good for stacking Flutter `Widget`s on-top, transparency, scrolling etc.

On the other hand, [flutter_native_view](https://github.com/alexmercerind/flutter_native_view) can already be used to embed XAML [`<WebView />`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webview?view=winrt-22000) "UWP" or [Edge WebView2](https://docs.microsoft.com/en-us/microsoft-edge/webview2/get-started/win32) (even easier for Win32) or [`<MediaPlayerElement />`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.mediaplayerelement?view=winrt-22000) "UWP" inside a Flutter app. [flutter_native_view](https://github.com/alexmercerind/flutter_native_view) is currently capable of:

- Opaque/Transparent/Translucent Flutter `Widget`s on top.
- Intractability with native surfaces.
- Scrollable `Widget`s like `ListView`.
- Stacking.
- Multiple instances of native surfaces.
- Other features that I worked on.

![yRRmPWZ8A7](https://user-images.githubusercontent.com/28951144/159650827-e4d346f5-8958-4c93-ab77-13ce1daf507e.gif)

[GIF showing a VLC instance (buttons can be disabled through libvlc API) & XAML [`<WebView />`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webview?view=winrt-22000) inside Flutter Win32 runner. Slight lag can be noticed due to screen-recording & other current inefficiencies.]

Though, [flutter_native_view](https://github.com/alexmercerind/flutter_native_view) may look tempting but REMEMBER currently it depends upon an undocumented DWM API (and not safe for usage in [flutter/engine](https://github.com/flutter/engine)). In future, I will try to switch to some different technique to allow transparent window backgrounds (other than `WS_EX_LAYERED`, which just allows masking of a `COLORKEY`).

There are certainly problems as of now, and will require few tweaks here & here to get it _just about right_ for usage.

**Comment by gslender** [comment]
Eagerly waiting for this 🙏

**Comment by mohamadlounnas** [comment]
any update?

**Comment by wanjm** [comment]
platfromView support on macos has not finished yet, and support on Windows will begin after that, does that means this issue will take another one year to fix.

**Comment by NickYoung-max** [comment]
flutter webview using webview_windows [https://pub.dev/packages/webview_windows](url)，html tag  like <select>...</select> is not work correctly,it's mostly because of the focus event.When mouse clicked, flutter view get focus first,then webview lose focus, and then flutter view lose focus and webview get focus finally.In this way ,<select> tag should pop up a menu but nothing happended.How can I do to make the <select>tag work collectly in flutter webview?

**Comment by RakshithSarja** [comment]
@stuartmorgan Could you please update when can we expect this feature support for windows?

**Comment by stuartmorgan** [comment]
@RakshithSarja Please see https://github.com/flutter/flutter/issues/31713#issuecomment-756740559

**Comment by wanjm** [comment]
we may wait for this to be resolved.https://github.com/flutter/flutter/issues/128904

**Comment by leisu1** [comment]
@stuartmorgan  Hi，my company‘s’ business requires a webview embedded in the main window. I tried to add a webview to the window of the platform, exchange data through channel, and control the display through hidden/show. This was successful on macos, but webview content was not displayed on windows. 
I wrote a simple demo verification, added a button, and moved flutterview to the right a little, the button was displayed on the window, but the part covered by flutterview could not be displayed, what is the reason? Is there a way to change it?
![屏幕截图 2023-12-28 193410](https://github.com/flutter/flutter/assets/19525360/ed4044ef-a237-4aed-b249-9d11ab17ff45)

**Comment by stuartmorgan** [comment]
@leisu1 Please see https://github.com/flutter/flutter/wiki/Issue-hygiene#comments-providing-workarounds

This issue is about implementing platform view support; discussions of how to implement your own alternate approaches to displaying native content are off-topic here.

**Comment by wanjm** [comment]
it was planned to implement in 2023 as road map shows, but it has not begin yet, where can we get the plan, we have been waitting for many years;

**Comment by stuartmorgan** [comment]
@wanjm I have previously (in this issue) pointed you to our policy against asking for updates on issue status. Please take a moment to review our code of conduct.

**Comment by wanjm** [comment]
@stuartmorgan  yes, but currently it is different, as it was in road map of 2023. so I ask again. sorry for my eager for it;

**Comment by DavidDohmen** [comment]
@stuartmorgan this has a business impact on our company. We'd be willing contribute financially but don't have the means to do so with developer time. Is there a bounty system in place for these kind of issues/feature developments? 
Since many companies are in similar situations like ours, this could potentially help funnel money and thus motivated indipendent open-source developers to the overall improvement of flutter. What do you think?

**Comment by stuartmorgan** [comment]
The Flutter team does not manage any kind of bounty system. Nothing would prevent interested parties from organizing their own bounties however they like, however.

**Comment by dahchon** [comment]
> The Flutter team does not manage any kind of bounty system. Nothing would prevent interested parties from organizing their own bounties using whatever system works for them, however.

have this changed? 
the roadmap 2024 (https://github.com/flutter/flutter/wiki/Roadmap) here mentioned platform views of windows / mac are your focus on 2024

**Comment by mustafahimi** [comment]
After 5 years, still nothing?

**Comment by Im-Kevin** [comment]
Platform View and Muilt Window have always been two pain points of Flutter desktop. If they can be solved, it will definitely have a place in desktop development and may even surpass Electron. I hope the official and community can move forward faster.

**Comment by Sylwester50** [comment]
any roadmap on that?
